src/toil/jobStores/fileJobStore.py

```diff
@@ -215,13 +221,55 @@ def getSize(cls, url):
 
     @classmethod
     def _readFromUrl(cls, url, writable):
-        with open(cls._extractPathFromUrl(url), 'r') as f:
-            writable.write(f.read())
+        """
+        Writes the contents of a file to a source (writes url to writable).
+
+        Uses a buffer of 1Gb if file is larger than 1Gb, else
+        read the entire file into memory for writing.
+
+        :param str url: A path as a string of the file to be read from.
+        :param object writable: An open file object to write to.
+        """
+        file_path = cls._extractPathFromUrl(url)
+        with open(file_path, 'rb') as f:
+            if os.path.getsize(file_path) > cls.SMALL_FILE_SIZE_THRESHOLD:
+                # We use a buffer here because otherwise the system attempts to read the
+                # entire file into RAM, and if the file is larger than the available RAM,
+                # it causes a MemoryError.
+                while True:
+                    data = f.read(cls.BUFFER_SIZE)
+                    if not data:
+                        break
+                    writable.write(data)
+            else:
+                # Using a buffer is slower (tested), so we only want to use it on large files
+                writable.write(f.read())
 
     @classmethod
     def _writeToUrl(cls, readable, url):
-        with open(cls._extractPathFromUrl(url), 'w') as f:
-            f.write(readable.read())
+        """
+        Writes the contents of a file to a source (writes readable to url).
+
+        Uses a buffer of 1Gb if file is larger than 1Gb, else
+        read the entire file into memory for writing.
+
+        :param str url: A path as a string of the file to be written to.
+        :param object readable: An open file object to read from.
+        """
+        file_path = cls._extractPathFromUrl(url)
+        with open(file_path, 'wb') as f:
+            if os.fstat(readable.fileno()).st_size > cls.SMALL_FILE_SIZE_THRESHOLD:
```

Comment:
Thanks for catching this!  Alright, I've implemented only using the buffer code.  Hannes also pointed out that shutil has a copyfileobj() function that uses buffering.  I did some speed tests that ran overnight and decided to set their buffer a bit higher to 10Mb, rather than their default of 16Kb.  Their source code is exactly the same as the buffering portion of the function I originally had, so calling it should be equivalent.  Thanks again Joel!

Reply:
Nice! I forgot all about `shutil.copyfileobj`. That's perfect for our use case.